chef/automate
# experiment: buf API linting

### :nut_and_bolt: Description: What code changed, and why?

I've been wanting to see what [buf](https://buf.build)'s checks say about our proto files. 

⚠️ The grpc.sh protoc arguments have not been updated here yet. This will break in many ways -- I've merely changed the proto file layout such that we are able to use `buf`'s checks.

### :athletic_shoe: How to Build and Test the Change

1. Download buf: https://github.com/bufbuild/buf/releases/tag/v0.13.0
2. Run its checks in the top-level of this branch: `buf check lint`
3. Create an image for the checkout: `buf image build -o 28c177197e0042c9dbdab05b3f53f0050923d8d8.json`
4. Change some things in some proto file, for example:
```diff
diff --git a/api/external/applications/applications.proto b/api/external/applications/applications.proto
index 3b1afee99..481ff93df 100644
--- a/api/external/applications/applications.proto
+++ b/api/external/applications/applications.proto
@@ -31,7 +31,7 @@ service ApplicationsService {
   */
   rpc GetServiceGroups(ServiceGroupsReq) returns (ServiceGroups) {
-    option (google.api.http).get = "/api/v0/applications/service-groups";
+    option (google.api.http).get = "/api/v1/applications/service-groups";
     option (chef.automate.api.iam.policy) = {
       resource: "applications:serviceGroups"
       action: "applications:serviceGroups:list"
@@ -391,14 +391,15 @@ message ServicesReq {
   common.query.Pagination pagination = 2;
   // Applies sorting parameters.
   common.query.Sorting sorting = 3;
+  string new_field = 4;
 }
 
 // Request message for listing available service filters.
 message ServicesDistinctValuesReq {
   // Field name of service values.
-  string field_name = 1;
+  string field_name = 2;
   // Query value, supports wildcards (* and ?).
-  string query_fragment = 2;
+  string query_fragment = 1;
   // Applies filters, in the format of `fieldname:value`.
   // See documentation for ServicesReq for valid filter parameters
   repeated string filter = 3;
@@ -644,6 +645,7 @@ enum HealthStatus {
   CRITICAL = 2;
   UNKNOWN = 3;
   NONE = 17;
+  ANOTHERONE = 200;
 }
 
 // HealthCheckResult aggregates the stdout output, stderr output and process
```

4. Run `buf check breaking --against-input 28c177197e0042c9dbdab05b3f53f0050923d8d8.json`:
```
api/external/applications/applications.proto:400:3:Field "2" with name "field_name" on message "ServicesDistinctValuesReq" changed option "json_name" from "queryFragment" to "fieldName".
api/external/applications/applications.proto:400:10:Field "2" on message "ServicesDistinctValuesReq" changed name from "query_fragment" to "field_name".
api/external/applications/applications.proto:402:3:Field "1" with name "query_fragment" on message "ServicesDistinctValuesReq" changed option "json_name" from "fieldName" to "queryFragment".
api/external/applications/applications.proto:402:10:Field "1" on message "ServicesDistinctValuesReq" changed name from "field_name" to "query_fragment".
```

#### Observations

- Breaking change detection meets expectations. 😉 
- Breaking changes to the `google.api.http` annotations are not captured -- I assume those are beyond buf's (current) scope
- The linting results looks much more useful than what we have so far (see below), like including good practices around package names, enum values.
- Both the breaking change check and the lint check are fast.

[:mag: Lint results](https://gist.github.com/srenatus/8a2faa240221f78d49c04e0748957ea3)
